game-core/src/main/java/games/strategy/triplea/ui/BattleDisplay.java

```diff
@@ -375,8 +367,33 @@ private Territory getSubmerge(final String message) {
     return retreatTo.get();
   }
 
-  private boolean showSubmergeDialog(
-      final String message, final AtomicReference<Territory> retreatTo) {
+  private Action getPlayerAction(
```

Comment:
This still feels a bit clunky - since it's not just some random player action, but something that returns a RetreatResult.

How about just not having this method and inlining it into getRetreat() instead?

e.g.

```
final String title;
final Suppler<RetreatResult> showDialog;
if (!submerge || possible.size() > 1) {
  title = "Retreat?";
  showDialog = () -> showSubmergeDialog(message), retreatTo, latch);
} else {
  title = "Submerge Subs?";
  showDialog = () -> showSubmergeDialog(message);
}
```

And then the inlined SwingAction.of() call.

Otherwise it's hard to follow what's going on especially since there is a lot of stuff happening in this class already - so adding all these new methods just introduces more complexity.

Another option is to move the logic into separate classes for the logic, but not sure of the best set up there...

... If we step back a bit, even with this refactoring, I feel that this classes suffers a lot from mixing game logic with Swing UI code. Ultimately we'd probably want to recode things such that low level concepts - like use of JOptionPane - are abstracted away, but the business logic remains. But that's probably too big of a change for what you were after here.

Another though - I wonder if the Java Future interface would be better suited here than the combination of CountDownLatch and AtomicReference. Doesn't Future provide exactly what we need?

Reply:
Using future is an excellent idea, I'll try that.